### PR TITLE
Create codeql-analysis.yml

### PR DESCRIPTION
**Pull request by AlirezaSohofi**
# Description

Introduce [code scanning alert](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/about-code-scanning#about-tools-for-code-scanning) for this repository that runs on every merge to the main, every PR that is to be merged to the main, and also on a weekly schedule.

Fixes # issue

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring including code style reformatting 
- [ ] Other (please describe):

# Checklist:

- [ ] I have read the [contributing guideline doc](https://squirrel-core.readthedocs.io/en/latest/usage/code_of_conduct.html) (external contributors only)
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have kept the PR small so that it can be easily reviewed  
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] All dependency changes have been reflected in the pip requirement files.

### CODE REVIEW

File: .github/workflows/codeql-analysis.yml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    # ℹ️ Command-line programs to run using the OS shell.
+    # 📚 See https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsrun
+
+    #   If the Autobuild fails above, remove it and uncomment the following three lines.
+    #   modify them (or add more) to build your code if your project, please refer to the EXAMPLE below for guidance.
+
+    # - run: |
+    #   echo "Run, Build Application using script"
+    #   ./location_of_script_within_repo/buildscript.sh
```
Comment (AlpAribal) [inline .github/workflows/codeql-analysis.yml:72]: I don't think we need this part

File: .github/workflows/codeql-analysis.yml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+# supported CodeQL languages.
+#
+name: "CodeQL"
+
+on:
+  push:
+    branches: [ "main" ]
+  pull_request:
+    # The branches below must be a subset of the branches above
+    branches: [ "main" ]
```
Comment (AlpAribal) [inline .github/workflows/codeql-analysis.yml:19]: We won't be checking PRs that merge into other PRs, right? Is that on purpose?

**Review by AlpAribal — CHANGES_REQUESTED** [review]
Nice addition!

File: .github/workflows/codeql-analysis.yml (hunk 2 above)
Comment (AlirezaSohofi) [reply .github/workflows/codeql-analysis.yml:19]: No only on main

**Comment by AlpAribal** [comment]
When I view the CodeQL check output, I see the following warning message:
```
An error occurred while trying to automatically install Python dependencies: Error: The process '/home/runner/work/_actions/github/codeql-action/v2/python-setup/auto_install_packages.py' failed with exit code 1
Please make sure any necessary dependencies are installed before calling the codeql-action/analyze step, and add a 'setup-python-dependencies: false' argument to this step to disable our automatic dependency installation and avoid this warning.
```
Could not find anything substantial about this warning though.

**Review by AlpAribal — COMMENTED** [review]
Now, codeql analysis succeeds without errors or warnings. If this is enough, then LGTM. A few points are not clear to me though:
1. How do I parse or see the analysis results? When I check the action output, I see neither any errors nor a "no errors found" text
2. Do we need to install the requirements before the analysis? If the analysis is done on the source code, then installing dependencies might not be necessary.

File: .github/workflows/codeql-analysis.yml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        # Learn more about CodeQL language support at https://aka.ms/codeql-docs/language-support
+
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v3
+
+    - name: Set up Python
+      uses: actions/setup-python@v3
+      with:
+        python-version: '3.x'
```
Comment (AlpAribal) [inline .github/workflows/codeql-analysis.yml:50]: We have `python_requires=">=3.8.0"`. Should we test with the smallest allowed version (3.8.x) or with the latest version (3.10.x)?

File: .github/workflows/codeql-analysis.yml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+# supported CodeQL languages.
+#
+name: "CodeQL"
+
+on:
+  push:
+    branches: [ "main" ]
+  pull_request:
+    # The branches below must be a subset of the branches above
+    branches: [ "main" ]
```
[the remark this answers is not in the document]
Comment (AlirezaSohofi) [reply .github/workflows/codeql-analysis.yml:19]: updated the description :+1:

File: .github/workflows/codeql-analysis.yml (hunk 3 above)
Comment (AlirezaSohofi) [reply .github/workflows/codeql-analysis.yml:50]: Let's leave it at `3.x`

**Review by AlpAribal — APPROVED** [review]
LGTM!

**Pull request merged**